pithos/plugins/notification_icon.py

```diff
@@ -12,57 +12,142 @@
+        # or if we have the appindicatorsupport GNOME Shell extension installed and enabled,
+        # if not we check to see if we're running in KDE Plasma. If we aren't running in KDE Plasma,
+        # have the indicator.application service or have the appindicatorsupport GNOME Shell extension
+        # installed and enabled we have to check to see if we're on Wayland to see if we can use Gtk.StatusIcon.
+        # It's all rather convoluted...
+        def check_for_x():
+            self.appindicator = None
+            display = self.window.props.screen.get_display()
+            if not type(display).__name__.endswith('X11Display'):
+                logging.debug('Notification icon requires X11 or AppIndicator Support')
+                self.prepare_complete(error='Notification icon requires X11 or AppIndicator Support')
+            else:
+                self.prepare_complete()
+
+        def init_appindicator():
+            self.ind = self.appindicator.Indicator.new(
+                "pithos-tray-icon",
+                "pithos-tray-icon",
+                self.appindicator.IndicatorCategory.APPLICATION_STATUS,
+            )
+
             self.prepare_complete()
 
-    def on_enable(self):
-        self.delete_callback_handle = self.window.connect("delete-event", self._toggle_visible)
-        self.state_callback_handle = self.window.connect("play-state-changed", self.play_state_changed)
-        self.song_callback_handle = self.window.connect("song-changed", self.song_changed)
+        def has_appindicator_shell_ext():
+            schema = Gio.SettingsSchemaSource.get_default()
+            valid_schema = schema.lookup('org.gnome.shell', True)
+            if valid_schema:
+                gnome_shell = Gio.Settings.new('org.gnome.shell')
+                enabled_extensions = gnome_shell.get_value('enabled-extensions')
+                if 'appindicatorsupport@rgcjonas.gmail.com' in enabled_extensions:
+                    return True
+            return False
+
+        def on_plasmashell_finish(source, result, data):
+            try:
+                plasmashell = Gio.DBusProxy.new_finish(result)
+            except GLib.Error as e:
+                logging.debug(e)
+                check_for_x()
+            else:
+                bus_owned = plasmashell.get_name_owner()
+                if bus_owned and self.appindicator:
+                    init_appindicator()
+                elif bus_owned and not self.appindicator:
+                    logging.debug('Notification icon requires AppIndicator on Plasma')
+                    self.prepare_complete(error='Notification icon requires AppIndicator on Plasma')
+                else:
+                    check_for_x()
+
+        def on_indicator_finish(source, result, data):
+            try:
+                indicator_service = Gio.DBusProxy.new_finish(result)
+            except GLib.Error as e:
+                logging.debug(e)
+                check_for_x()
+            else:
```

Comment:
Could you sprinkle a few logging.debug's through-out here, it is sort of a complex startup trying a few different methods and it would be useful.